Color picker improvements, first round

afercia opened the pull request:
This PR tries to address a first round of improvements to the Color picker buttons.

- changes  aria-current to aria-pressed, see screenshot below to check how this the way buttons and itneraction with them get announced
- fix invalid HTML: buttons can't contain divs
- add missing aria-label on the "remove color" button
- make the Custom color picker toggle a real toggle that opens/closes the popover and uses aria-expanded
- fix the focus style on the Custom color picker: adds an inner span and applies overflow: hidden on it instead of applying it on the button (I guess this button with gradients needs to be tested a bit in IE 11 / Edge)

Fixes #2679

afercia committed:
Use aria-pressed, fix valid code, add missing aria-label.

afercia committed:
Properly toggle the Custom color picker.

afercia committed:
Fix focus style on the Custom color picker toggle.

afercia committed:
Properly rename call to toggle picker function.

afercia commented:
Screenshots to illustrate the interaction with Safari + VoiceOver:

![screen shot 2017-09-06 at 08 54 44](https://user-images.githubusercontent.com/1682452/30098343-d80a7252-92e1-11e7-8d1b-2d3b36e83c7c.png)

when clicking:

![screen shot 2017-09-06 at 08 54 49](https://user-images.githubusercontent.com/1682452/30098342-d8094e9a-92e1-11e7-94e9-244a7782255a.png)

when clicking again:

![screen shot 2017-09-06 at 08 54 53](https://user-images.githubusercontent.com/1682452/30098341-d80931bc-92e1-11e7-8ea6-8833fdec8a1f.png)

selected state:

![screen shot 2017-09-06 at 09 01 49](https://user-images.githubusercontent.com/1682452/30098399-15211100-92e2-11e7-8062-be8c4b4dd317.png)

aduth commented on blocks/color-palette/index.js:
```diff
@@ -42,38 +55,43 @@ class ColorPalette extends Component {
 					<Popover
 						isOpen={ this.state.opened }
-						onClose={ this.closePicker }
+						onClose={ this.closeOnClickOutside }
```
As of deb86a78f14196dfe20fa5297a3a7970dd82a8b8, `Popover`'s `onClose` callback no longer passes the click `event`, since there can be many reasons a Popover might emit a close intent. The change here should be simply to change `onClose={` to `onClickOutside={`

I expect you may need to `git rebase origin/master` and force push the branch for this to take effect, since #2697 was merged to master after your branch was created.

aduth commented on blocks/color-palette/index.js:
```diff
@@ -42,38 +55,43 @@ class ColorPalette extends Component {
+						ref={ this.bindToggleNode }
+						aria-label={ __( 'Custom color picker' ) }
+					>
+						<span className="blocks-color-palette__custom-color-gradient"></span>
```
Minor: For brevity, React allows you to self-close tags: `<span />` (**Edit:** I see this was written as such below, so more for consistency).

aduth commented on blocks/color-palette/index.js:
```diff
@@ -42,38 +55,43 @@ class ColorPalette extends Component {
 								onClick={ () => onChange( value === color ? undefined : color ) }
 								aria-label={ sprintf( __( 'Color: %s' ), color ) }
-								aria-current={ value === color && __( 'Selected color' ) }
+								aria-pressed={ value === color ? true : false }
```
The ternary shouldn't be necessary here, and can be simplified to `value === color` (will always be `true` or `false`)

aduth commented on blocks/color-palette/index.js:
```diff
@@ -42,38 +55,43 @@ class ColorPalette extends Component {
 						isOpen={ this.state.opened }
-						onClose={ this.closePicker }
+						onClose={ this.closeOnClickOutside }
+						onClick={ this.stopPropagation }
```
Can you explain the need for stopping propagation here?

afercia replied:
Hm not 100% sure given the time passed, I _think_ clicking anywhere on the ChromePicker (custom color) closed the popover but seems something has changed in the meantime (can't confirm). Will remove.

aduth reviewed (commented):
Apologies on the delay in reviewing this one.

afercia committed:
Process CR.

afercia commented:
Updated and rebased. But I'm seeing a _new_ CSS glitch in Safari 11, where the rounded focus style on the color buttons is not equally centered. Investigating.

<img width="303" alt="screen shot 2017-09-29 at 15 54 19" src="https://user-images.githubusercontent.com/1682452/31019067-c645dbb0-a52e-11e7-9b11-c40ba7c07dfb.png">

afercia committed:
Improve color buttons vertical alignment.

afercia committed:
Make ESLint happy.

afercia commented:
@aduth any specific reason for this rule:
`error  A space is required before closing bracket  react/jsx-space-before-closing`

seems a bit weird to me this actually can make any relevant difference:
```
<span className="blocks-color-palette__custom-color-gradient"/>
<span className="blocks-color-palette__custom-color-gradient" />
```

afercia commented:
The alignment issue in Safari 11 is solved in https://github.com/WordPress/gutenberg/pull/2680/commits/05038752844ff48e42e2a560ba4896f6f9f3bdbc. Any objections to merge?

aduth commented:
>@aduth any specific reason for this rule:

Like rules on spacing within parentheses, it's merely a style convention aimed at encouraging consistency.

I would be glad to add these to the coding guidelines document but... the current atmosphere leads me to believe that it would be unwise to be codifying our React usage patterns at this point in time 😬 

(Will follow-up with a final review)

aduth committed:
Remove unused stopPropagation method

aduth reviewed (approved):
Removed unused `stopPropagation` method in a4d10f7. Otherwise LGTM 👍

aduth commented:
Also, in case you're not already using one and are interested, there are a number of editor integrations to surface ESLint issues before it reaches CI: https://eslint.org/docs/user-guide/integrations

afercia commented:
thanks!
